openshift-kube-apiserver/admission/customresourcevalidation/imagecontentsourcepolicy/validate_imagecontentsourcepolicy.go

```diff
@@ -0,0 +1,131 @@
+	itmscli cfgv1client.ImageTagMirrorSetsGetter
+}
+
+func newValidateICSP() (admission.Interface, error) {
+	ret := &validateCustomResourceWithClient{}
+
+	delegate, err := customresourcevalidation.NewValidator(
+		map[schema.GroupResource]bool{
+			v1alpha1.Resource(imagedigestmirrorset.IcspPlural): true,
+		},
+		map[schema.GroupVersionKind]customresourcevalidation.ObjectValidator{
+			v1alpha1.GroupVersion.WithKind(imagedigestmirrorset.IcspKind): imagecontentsourcepolicyV1alpha1{idmscli: ret.getIdmscli, itmscli: ret.getItmscli},
+		})
+	if err != nil {
+		return nil, err
+	}
+	ret.ValidationInterface = delegate
+	return ret, nil
+}
+
+var _ admissionrestconfig.WantsRESTClientConfig = &validateCustomResourceWithClient{}
+
+func (i *validateCustomResourceWithClient) getIdmscli() cfgv1client.ImageDigestMirrorSetsGetter {
+	return i.idmscli
+}
+
+func (i *validateCustomResourceWithClient) getItmscli() cfgv1client.ImageTagMirrorSetsGetter {
+	return i.itmscli
+}
+
+func (i *validateCustomResourceWithClient) SetRESTClientConfig(restClientConfig rest.Config) {
+	cli, err := cfgv1client.NewForConfig(&restClientConfig)
+	if err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
+	i.idmscli = cli
+	i.itmscli = cli
+}
+
+func (i *validateCustomResourceWithClient) ValidateInitialization() error {
+	if i.idmscli == nil {
+		return fmt.Errorf(PluginName + " needs an imageDigestMirrorSetsGetter")
+	}
+
+	if i.itmscli == nil {
+		return fmt.Errorf(PluginName + " needs an imageTagMirrorSetsGetter")
+	}
+
+	if initializationValidator, ok := i.ValidationInterface.(admission.InitializationValidator); ok {
+		return initializationValidator.ValidateInitialization()
+	}
+
+	return nil
+}
+
+type imagecontentsourcepolicyV1alpha1 struct {
+	idmscli func() cfgv1client.ImageDigestMirrorSetsGetter
+	itmscli func() cfgv1client.ImageTagMirrorSetsGetter
+}
+
+func (i imagecontentsourcepolicyV1alpha1) ValidateCreate(uncastObj runtime.Object) field.ErrorList {
+	return i.validateICSPUse("create")
+}
+
+func (i imagecontentsourcepolicyV1alpha1) ValidateUpdate(uncastObj runtime.Object, uncastOldObj runtime.Object) field.ErrorList {
+	return i.validateICSPUse("update")
+}
+
+func (i imagecontentsourcepolicyV1alpha1) ValidateStatusUpdate(uncastObj runtime.Object, uncastOldObj runtime.Object) field.ErrorList {
+	return i.validateICSPUse("update")
+}
+
+func (i imagecontentsourcepolicyV1alpha1) validateICSPUse(action string) field.ErrorList {
+	idmsList, err := i.idmscli().ImageDigestMirrorSets().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		errs := field.ErrorList{field.InternalError(field.NewPath("Kind", imagedigestmirrorset.IdmsKind), err)}
+		return errs
+	}
+	if len(idmsList.Items) > 0 {
+		errs := field.ErrorList{
+			field.Invalid(field.NewPath("Kind", imagedigestmirrorset.IcspKind), "", fmt.Sprintf("can't %s %s when %s resources exist", action, imagedigestmirrorset.IcspKind, imagedigestmirrorset.IdmsKind))}
```

Comment:
Maybe `field.Forbidden` will much better express this fact, especially that it won't force you to put a value in there, but only reasons.

Reply:
Also if you're creating a variable just to return it, I'd suggest writing it as:
```golang
return field.ErrorList{
			field.Invalid(field.NewPath("Kind", imagedigestmirrorset.IcspKind), "", 
            fmt.Sprintf("can't %s %s when %s resources exist", action, imagedigestmirrorset.IcspKind, imagedigestmirrorset.IdmsKind))}
```
This applies to all error reported in here.